crates/core/component/shielded-pool/src/component/transfer.rs

```diff
@@ -540,14 +541,30 @@ impl AppHandlerExecute for Ics20Transfer {
 
     async fn timeout_packet_execute<S: StateWrite>(mut state: S, msg: &MsgTimeout) -> Result<()> {
         // timeouts may fail due to counterparty chains sending transfers of u128-1
-        timeout_packet_inner(&mut state, msg)
+        timeout_packet_inner(&mut state, &msg.packet)
             .await
             .context("able to timeout packet")?;
 
         Ok(())
     }
 
-    async fn acknowledge_packet_execute<S: StateWrite>(_state: S, _msg: &MsgAcknowledgement) {}
+    async fn acknowledge_packet_execute<S: StateWrite>(
+        mut state: S,
+        msg: &MsgAcknowledgement,
+    ) -> Result<()> {
+        let ack: TokenTransferAcknowledgement =
+            serde_json::from_slice(msg.acknowledgement.as_slice())?;
+        if !ack.is_successful() {
+            // in the case where a counterparty chain acknowledges a packet with an error,
+            // for example due to a middleware processing issue or other behavior,
+            // the funds should be unescrowed back to the packet sender.
+            timeout_packet_inner(&mut state, &msg.packet)
```

Comment:
So the logic in the ics 20 specification for handling error acknowledgements and for handling timeouts appears identical. 

I do wonder if it would be more maintainable to make the code a bit more like spec.

Otherwise looks good

Reply:
@avahowell can comment more here but I think the goal was to have a minimal diff that used the existing (audited) code as much as possible

Reply:
Yes, I checked and verified that the behavior should be identical, then kept the diff minimal in order to minimize risk and maximize readability